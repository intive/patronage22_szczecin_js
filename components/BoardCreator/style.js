import styled from 'styled-components'
import Button from '../Button/Button'
import { breakpoint } from '../../config/breakpoints'

export const Modal = styled.div`
  position: fixed;
  z-index: 10;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  display: flex;
  flex-direction: column;
  background: #FFFFFF;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  background: #FFFFFF;
  overflow: hidden;
`

export const CloseButton = styled(Button)`
  position: fixed;
  top: 1rem;
  right: 0rem;
  box-shadow: none;
  cursor: pointer;
  background: transparent;
  color: #000000;

  &:hover {
    background: transparent;
    color: #7E7E7E;
  }

  @media only screen and ${breakpoint.md} {
    top: 1.5rem;
    right: 2.5rem;
  }
`

export const BoardContent = styled.div`
  position: relative;
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 100%;
  padding: 0rem 1rem;
  background: #FFFFFF;

  @media only screen and ${breakpoint.md} {
    width: 34rem;
    height: 20.25rem;
    padding: 0rem;
  }
`

export const Title = styled.div`
  font-size: 1.5rem;
  line-height: 2rem;
  text-align: center;
  color: #515151;
`

export const StyledButton = styled(Button)`
  position: fixed;
  bottom: 0;
  width: 90%;
  margin: 1rem;

  @media only screen and ${breakpoint.md} {
    position: absolute;
    bottom: 0;
    right: 0;
    width: auto;
    margin: 0rem;
  }
`
