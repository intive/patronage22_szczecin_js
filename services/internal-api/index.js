import axios from 'axios'

const axiosInstance = axios.create({
  baseURL: `https://${process.env.NEXT_PUBLIC_VERCEL_URL}` || 'http://localhost:3000',
  responseType: 'json',
  withCredentials: true
})

export const getBoards = async (...args) => {
  return axiosInstance.get('/api/boards', { params: Object.assign({}, ...args) })
}
