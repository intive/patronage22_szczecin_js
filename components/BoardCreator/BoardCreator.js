import { CloseButton, Modal, Title, BoardContent, StyledButton } from './style'
import { useTranslation } from 'next-i18next'
import BoardCreatorWizard from '../BoardCreatorWizard/BoardCreatorWizard'
import ReactPortal from '../ReactPortal/ReactPortal'
import { useEffect, useState, useContext } from 'react'
import { addBoard } from '../../services/internal-api'
import BoardsContext from '../../store/boards-context'

const BoardCreator = ({ isOpen, onClose }) => {
  const { t } = useTranslation('common')
  const [boardName, setBoardName] = useState('')
  const [isModalOpen, setIsModalOpen] = useState(isOpen)
  const boardsCtx = useContext(BoardsContext)
  const properties = { minLength: 5, maxLength: 50 }

  const handleOnChange = (event) => {
    setBoardName(event.target.value)
  }

  const closeModal = () => {
    setIsModalOpen(false)
    try {
      onClose()
    } catch (err) {}
  }

  const addBoardHandler = async (boardName) => {
    closeModal()
    try {
      const data = await addBoard(boardName)
      if (data.status === 204) {
        boardsCtx.reload()
      }
    } catch (err) {}
  }

  useEffect(() => {
    setBoardName('')
    setIsModalOpen(isOpen)
  }, [isOpen])

  return isModalOpen
    ? (
      <ReactPortal wrapperId='modal-portal'>
        <Modal aria-hidden='false' role='dialog'>
          <CloseButton onClick={closeModal} icon='close' />
          <BoardContent>
            <Title>{t('createNewBoard.createNewBoardTitle')}</Title>
            <BoardCreatorWizard onChange={handleOnChange} value={boardName} maxLength={properties.maxLength} />
            <StyledButton disabled={boardName.trim().length < properties.minLength} onClick={() => addBoardHandler(boardName.trim())}>{t('buttons.continue')}</StyledButton>
          </BoardContent>
        </Modal>
      </ReactPortal>
      )
    : null
}

export default BoardCreator
