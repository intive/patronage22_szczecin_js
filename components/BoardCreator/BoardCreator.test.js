/* globals describe, it, expect  */
/* eslint-env jest */

import { render, screen, waitForElementToBeRemoved } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import BoardCreator from './BoardCreator'
import * as InternalApi from '../../services/internal-api'

describe('BoardCreator', () => {
  it('renders modal BoardCreator', () => {
    const { baseElement } = render(
      <BoardCreator isOpen />
    )

    expect(baseElement).toMatchSnapshot()
  })

  it('should close modal after click on continue button', async () => {
    render(<BoardCreator isOpen />)
    userEvent.type(screen.getByRole('textbox'), 'Test5')
    const button = screen.getByRole('button', { name: /buttons.continue/ })
    userEvent.click(button)

    waitForElementToBeRemoved(button).then(() => expect(button).not.toBeInTheDocument())
  })

  it('should close modal after click on cross button', () => {
    render(<BoardCreator isOpen />)
    const button = screen.getByRole('button', { name: /close/ })
    userEvent.click(button)
    expect(button).not.toBeInTheDocument()
  })

  it('should filled input field correctly', () => {
    const inputData = { boardname: 'My new board' }
    const { baseElement } = render(
      <BoardCreator isOpen />
    )
    const boardnameInput = screen.getByRole('textbox')
    boardnameInput.value = inputData.boardname

    expect(boardnameInput.value.length).toBe(12)
    expect(boardnameInput.value).toBe('My new board')
    expect(baseElement).toMatchSnapshot()
  })

  it('should show the continue button with the attribute disabled after rendering BoardCreator', () => {
    render(<BoardCreator isOpen />)

    expect(screen.getByText('buttons.continue')).toHaveAttribute('disabled')
  })

  it('should remove disbaled atribute on continue button after type 5 characters in text area', () => {
    render(<BoardCreator isOpen />)

    expect(screen.getByText('buttons.continue')).toHaveAttribute('disabled')

    userEvent.type(screen.getByRole('textbox'), 'Test5')

    expect(screen.getByRole('textbox')).toHaveValue('Test5')
    expect(screen.queryByText('buttons.continue')).not.toHaveAttribute('disabled')
  })

  it('should call add option', () => {
    const mockAddBoard = jest.spyOn(InternalApi, 'addBoard')
    render(<BoardCreator isOpen />)
    userEvent.type(screen.getByRole('textbox'), 'Test5')
    userEvent.click(screen.queryByText('buttons.continue'))

    expect(mockAddBoard).toHaveBeenCalledTimes(1)
  })
})
