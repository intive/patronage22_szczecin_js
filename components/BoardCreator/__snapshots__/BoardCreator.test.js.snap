// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardCreator renders modal BoardCreator 1`] = `
<body>
  <div />
  <div
    id="modal-portal"
  >
    <div
      aria-hidden="false"
      class="sc-dkPtRN jzSsRw"
      role="dialog"
    >
      <button
        class="sc-bdvvtL dRBcfm sc-hKwDye kMksWp"
      >
        <span
          class="material-icons sc-gsDKAQ jUyrhl"
        >
          close
        </span>
      </button>
      <div
        class="sc-eCImPb brrNer"
      >
        <div
          class="sc-jRQBWg tHMPR"
        >
          createNewBoard.createNewBoardTitle
        </div>
        <div
          class="sc-iCfMLu ftPdqH"
        >
          createNewBoard.createNewBoardSubtitle
        </div>
        <input
          class="sc-furwcr jnaAwp"
          type="text"
        />
        <button
          class="sc-bdvvtL dRBcfm sc-gKclnd eWCgBz"
        >
          buttons.continue
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`BoardCreator should close modal after click on continue button 1`] = `
<body>
  <div />
  <div
    id="modal-portal"
  >
    <div
      aria-hidden="false"
      class="sc-dkPtRN jzSsRw"
      role="dialog"
    >
      <button
        class="sc-bdvvtL dRBcfm sc-hKwDye kMksWp"
      >
        <span
          class="material-icons sc-gsDKAQ jUyrhl"
        >
          close
        </span>
      </button>
      <div
        class="sc-eCImPb brrNer"
      >
        <div
          class="sc-jRQBWg tHMPR"
        >
          createNewBoard.createNewBoardTitle
        </div>
        <div
          class="sc-iCfMLu ftPdqH"
        >
          createNewBoard.createNewBoardSubtitle
        </div>
        <input
          class="sc-furwcr jnaAwp"
          type="text"
        />
        <button
          class="sc-bdvvtL dRBcfm sc-gKclnd eWCgBz"
        >
          buttons.continue
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`BoardCreator should close modal after click on cross button 1`] = `
<body>
  <div />
  <div
    id="modal-portal"
  >
    <div
      aria-hidden="false"
      class="sc-dkPtRN jzSsRw"
      role="dialog"
    >
      <button
        class="sc-bdvvtL dRBcfm sc-hKwDye kMksWp"
      >
        <span
          class="material-icons sc-gsDKAQ jUyrhl"
        >
          close
        </span>
      </button>
      <div
        class="sc-eCImPb brrNer"
      >
        <div
          class="sc-jRQBWg tHMPR"
        >
          createNewBoard.createNewBoardTitle
        </div>
        <div
          class="sc-iCfMLu ftPdqH"
        >
          createNewBoard.createNewBoardSubtitle
        </div>
        <input
          class="sc-furwcr jnaAwp"
          type="text"
        />
        <button
          class="sc-bdvvtL dRBcfm sc-gKclnd eWCgBz"
        >
          buttons.continue
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`BoardCreator should filled input field correctly 1`] = `
<body>
  <div />
  <div
    id="modal-portal"
  >
    <div
      aria-hidden="false"
      class="sc-dkPtRN jzSsRw"
      role="dialog"
    >
      <button
        class="sc-bdvvtL dRBcfm sc-hKwDye kMksWp"
      >
        <span
          class="material-icons sc-gsDKAQ jUyrhl"
        >
          close
        </span>
      </button>
      <div
        class="sc-eCImPb brrNer"
      >
        <div
          class="sc-jRQBWg tHMPR"
        >
          createNewBoard.createNewBoardTitle
        </div>
        <div
          class="sc-iCfMLu ftPdqH"
        >
          createNewBoard.createNewBoardSubtitle
        </div>
        <input
          class="sc-furwcr jnaAwp"
          type="text"
        />
        <button
          class="sc-bdvvtL dRBcfm sc-gKclnd eWCgBz"
        >
          buttons.continue
        </button>
      </div>
    </div>
  </div>
</body>
`;
